### PR TITLE
Feature/auto select dataprovider source

### PR DESCRIPTION
**Pull request by deanhannigan**
## Description

Data providers will now automatically select a datasource when added to a screen for the first time. 

The order of preference for datasource auto-selection is as follows:

1. Test DB (https://github.com/Budibase/budibase/issues/8308)
2. Budibase DB (Excludes the Users table)
3. External Source. (MySQL, Postgres etc.)

When the component instance is created, the first table in the selected datasource will appear in the provider.

Addresses: 
- https://github.com/Budibase/budibase/issues/8309

## Screenshots

Components like the Table block will load immediately with the pre-determined datasource.

output

### CODE REVIEW

File: packages/builder/src/builderStore/store/frontend.js [hunk 1]
```diff
@@ -481,6 +481,57 @@ export const getFrontendStore = () => {
           return null
         }
 
+        let dataSourceField = definition.settings.find(
```
Comment (aptkingston) [inline packages/builder/src/builderStore/store/frontend.js:484]: For this first half - we actually have a util called `getComponentSettings` in `componentUtils.js` which will get the full list of settings for  component type, including flattening and merging settings inside section. We could use that here to reduce a lot of this code!

File: packages/builder/src/builderStore/store/frontend.js [hunk 2]
```diff
@@ -490,6 +541,15 @@ export const getFrontendStore = () => {
           }
         })
 
+        // Set a default datasource
+        if (dataSourceField && defaultDatasource) {
+          props["dataSource"] = {
```
Comment (aptkingston) [inline packages/builder/src/builderStore/store/frontend.js:546]: We can't assume the name of the setting is `dataSource` here - we should reference the `key` property of the datasource/table setting found earlier.

**Review by aptkingston — COMMENTED** [review]
Couple of small suggestions to refactor but looks good! Great QoL improvement.

File: packages/builder/src/builderStore/store/frontend.js [hunk 3]
```diff
@@ -481,15 +481,60 @@ export const getFrontendStore = () => {
...
+        let defaultDatasource
+        if (dataSourceField) {
+          const _tables = get(tables)
+          const filteredTables = _tables.list.filter(
+            table => table._id != "ta_users"
+          )
+
+          const internalTable = filteredTables.find(
+            table =>
+              table.sourceId === "bb_internal" && table.type == "internal"
```
Comment (shogunpurple) [inline packages/builder/src/builderStore/store/frontend.js:500]: NAB: Might be nice to use constants here

File: packages/builder/src/builderStore/store/frontend.js [hunk 4]
```diff
@@ -481,15 +481,60 @@ export const getFrontendStore = () => {
...
+          const defaultSourceTable = filteredTables.find(
+            table =>
+              table.sourceId !== "bb_internal" && table.type == "internal"
+          )
+
+          const defaultExternalTable = filteredTables.find(
+            table => table.type == "external"
+          )
+
+          if (defaultSourceTable) {
```
Comment (shogunpurple) [inline packages/builder/src/builderStore/store/frontend.js:512]: Possible simplification:

`const defaultDatasource = defaultSourceTable || internalTable || defaultExternalTable`

**Review by shogunpurple — APPROVED** [review]
LGTM! Optional comments.

**Pull request merged**